src/backend/distributed/commands/schema_based_sharding.c

```diff
@@ -227,18 +575,230 @@ citus_internal_unregister_tenant_schema_globally(PG_FUNCTION_ARGS)
+
+	Oid schemaId = PG_GETARG_OID(0);
+
+	/* Prevent concurrent schema deletion and table creation under the schema */
+	LockDatabaseObject(NamespaceRelationId, schemaId, 0, AccessExclusiveLock);
+
+	/*
+	 * It is possible that by the time we acquire the lock on schema,
+	 * concurrent DDL has removed it. We can test this by checking the
+	 * existence of schema.
+	 */
+	if (!SearchSysCacheExists1(NAMESPACEOID, ObjectIdGetDatum(schemaId)))
+	{
+		ereport(ERROR, (errcode(ERRCODE_UNDEFINED_SCHEMA),
+						errmsg("concurrent operation removed the schema")));
+	}
+
+	/* Ensure schema checks */
+	EnsureSchemaOwner(schemaId);
+
+	/* The schema should be a tenant schema */
+	char *schemaName = get_namespace_name(schemaId);
+	if (!IsTenantSchema(schemaId))
+	{
+		ereport(ERROR, (errmsg("schema %s is not a tenant schema", schemaName)));
+	}
+
+	ereport(NOTICE, (errmsg("converting the schema %s back to a regular schema",
+							schemaName)));
+
+	/* Only single shard tables are expected during the undistribution of the schema */
+	List *disallowedCitusTableTypesForUndistribute PG_USED_FOR_ASSERTS_ONLY =
+		DisallowedTableTypesForUndistributeSchema();
+	SchemaTables *schemaTables PG_USED_FOR_ASSERTS_ONLY = TablesUnderSchema(schemaId);
+	Assert(!SchemaHasCitusTableType(schemaTables,
+									disallowedCitusTableTypesForUndistribute));
+	Assert(schemaTables->postgresTableIdList == NIL);
+
+	/*
+	 * First, we need to delete schema metadata and sync it to workers. Otherwise,
+	 * we would get error from `ErrorIfTenantTable` while undistributing the tables.
+	 */
+	char *unregisterSchemaCommand = TenantSchemaDeleteCommand(schemaName);
+	DeleteTenantSchemaLocally(schemaId);
+	if (EnableMetadataSync)
+	{
+		SendCommandToWorkersWithMetadata(unregisterSchemaCommand);
+	}
+
+	/* Undistribute all single shard tables in the schema */
+	List *tablesToConvert = NIL;
+	Oid relationId = InvalidOid;
+	foreach_oid(relationId, schemaTables->singleShardTableIdList)
+	{
+		/*
+		 * Skip partitions as they would be undistributed by the parent table.
+		 *
+		 * We should filter out partitions here before undistributing the schema.
+		 * Otherwise, converted partitioned table would change oid of partitions and its
+		 * partition tables would fail with oid not exist.
+		 */
+		if (PartitionTable(relationId))
+		{
+			continue;
+		}
+
+		/*
+		 * Error early before undistributing the schema if concurrent drop or some alter
+		 * statements occurred on the table. Note that we already had the AccessShareLock
+		 * lock when we call PartitionTable, so after this point we are guaranteed to
+		 * prevent concurrent drop and some modification on the table.
+		 */
+		EnsureRelationExists(relationId);
+		tablesToConvert = lappend_oid(tablesToConvert, relationId);
+	}
+
+	UndistributeTables(tablesToConvert);
+
+	/* Delete colocation metadata and sync to workers */
+	uint32 schemaColocationId = SchemaIdGetTenantColocationId(schemaId);
+	DeleteColocationGroup(schemaColocationId);
```

Comment:
Since we delete the tenant schema record from the metadata earlier in this function, we can simply drop those three lines because the colocation group would automatically get dropped after all the tables in the schema are undistributed.

Even more, rn we're probably trying to delete a colocation group with INVALID_COLOCATION_ID.

In other words, we automatically delete the colocation groups when they become an empty one, unless the tenant schema is used by a tenant schema. However, given that we've already deleted the entry related to tenant schema from metadata earlier this function, UndistributeTables should automatically be deleting the colocation group already.